pkg/operator/ceph/controllers/clusterdisruption/location.go

```diff
@@ -0,0 +1,216 @@
+		osdID, ok := labels[osd.OsdIdLabelKey]
+		if !ok {
+			return nil, fmt.Errorf("osd %s was not labeled with ", deployment.ObjectMeta.Name)
+		}
+		osdIDInt, err := strconv.Atoi(osdID)
+		if err != nil {
+			return nil, fmt.Errorf("osd id not an int: %s", osdID)
+		}
+		crushMeta, err := r.osdCrushLocationMap.Get(request.Namespace, osdIDInt)
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch info from ceph for osd %s", osdID)
+		}
+		osdData.CrushMeta = crushMeta
+		osds = append(osds, osdData)
+
+	}
+	return osds, nil
+}
+
+// OsdData stores the deployment and the Crush Data of the osd together
+type OsdData struct {
+	Deployment appsv1.Deployment
+	CrushMeta  *cephClient.CrushFindResult
+}
+
+// OSDCrushLocationMap is used to maintain a cache of map of osd id to cephClientCrushhFindResults
+// the crush position of osds wrt to the failureDomain is not expected to change often, so a default Resync period
+// of half an hour is used, but if a use case arises where this is needed, ResyncPeriod should be made smaller.
+type OSDCrushLocationMap struct {
+	ResyncPeriod       time.Duration
+	Context            *clusterd.Context
+	clusterLocationMap map[string]map[int]cachedOSDLocation
+	mux                sync.Mutex
+}
+
+type cachedOSDLocation struct {
+	result     *cephClient.CrushFindResult
+	lastSynced time.Time
+}
+
+// totest
+//Get takes an osd id and returns a CrushFindResult from cache
+func (o *OSDCrushLocationMap) Get(clusterNamespace string, id int) (*cephClient.CrushFindResult, error) {
+	o.mux.Lock()
+	defer o.mux.Unlock()
+	if o.ResyncPeriod == 0 {
+		o.ResyncPeriod = 30 * time.Minute
+	}
+
+	// initialize clusterLocationMap
+	if len(o.clusterLocationMap) == 0 {
+		o.clusterLocationMap = make(map[string]map[int]cachedOSDLocation)
+	}
+	locationMap, ok := o.clusterLocationMap[clusterNamespace]
+	// initialize namespace map
+	if !ok {
+		o.clusterLocationMap[clusterNamespace] = make(map[int]cachedOSDLocation)
+	}
+
+	// sync of osd id not found in clusterNamespace
+	osdLocation, ok := locationMap[id]
+	if !ok {
+		osdResult, err := o.get(clusterNamespace, id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to run `find` on osd %d in cluster %s", id, clusterNamespace)
+		}
+		o.clusterLocationMap[clusterNamespace][id] = cachedOSDLocation{result: osdResult, lastSynced: time.Now()}
+		return osdResult, nil
+	}
+
+	// sync if not synced for longer than ResyncPeriod
+	if time.Since(osdLocation.lastSynced) > o.ResyncPeriod {
+		osdResult, err := o.get(clusterNamespace, id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to run `find` on osd %d in cluster %s", id, clusterNamespace)
+		}
+		o.clusterLocationMap[clusterNamespace][id] = cachedOSDLocation{result: osdResult, lastSynced: time.Now()}
+		return osdResult, nil
+	}
+
+	return osdLocation.result, nil
+
+}
+
+// uncached version
+func (o *OSDCrushLocationMap) get(clusterNamespace string, id int) (*cephClient.CrushFindResult, error) {
```

Comment:
Why is this needed if we have a cached version?

Reply:
The cached version calls this.